### DISCUSSION TITLE
Google Search

### DISCUSSION THREAD

**Discussion by mabry1985** [johnlindquist/kit#1225] [Share]
I know this is redundant and might not be useful to many, but I needed to build a custom search tool for an agent I'm working on. 

I set this up to test the functionality and figured someone might find it useful.

```
/* 
# Google Search
Example of leveraging Google's Custom Search Engine API to search the web
*/

// Name: Google Search
// Description: Leverage Google's Custom Search Engine API to search the web
// Author: Josh Mabry
// Twitter: @AI_Citizen

import "@johnlindquist/kit";

let GOOGLE_API_KEY = await env("GOOGLE_API_KEY", {
  shortcuts: [
    {
      name: "Google API Key",
      key: `${cmd}+o`,
      bar: "right",
      onPress: () => {
        open("https://developers.google.com/custom-search/v1/introduction");
      },
    },
  ],
  ignoreBlur: true,
  secret: true,
  height: PROMPT.HEIGHT.INPUT_ONLY,
});

let GOOGLE_CSE_KEY = await env("GOOGLE_CSE_KEY", {
  shortcuts: [
    {
      name: "Google Custom Search Engine Key",
      key: `${cmd}+o`,
      bar: "right",
      onPress: () => {
        open("https://programmablesearchengine.google.com/");
      },
    },
  ],
  ignoreBlur: true,
  secret: true,
  height: PROMPT.HEIGHT.INPUT_ONLY,
});

let query = await arg(
  {
    placeholder: "Search Query",
    strict: false,
  },
  [
    {
      name: "Send a search query to Google",
      info: "always",
    },
  ]
);

let search = (q) =>
  `https://www.googleapis.com/customsearch/v1?key=${GOOGLE_API_KEY}&cx=${GOOGLE_CSE_KEY}&q=${q}&sort=date`;

let response = await get(search(query));

let items = response?.data?.items;

if (items) {
  let choices = items.map((item) => ({
    name: item.title,
    value: item.link,
  }));

  let link = await arg("Choose a link to view", choices);

  open(link);
}
```

**Comment by mabry1985** [comment]
always open to suggestions for improvement

**Reply by alwinraju** [reply]
Are you able to make one that uses SERP API? And then use's langchain's chatgpt model too? Similar to pairing tools in langchain?

**Reply by mabry1985** [reply]
@alwinraju absolutely! There is a native wrapper around the SERP [API](https://js.langchain.com/docs/api/tools/classes/SerpAPI) already actually. 

I just really don't like their pricing structure. Unless I'm missing something, it's 100 calls a month for the free tier, and then the next step up is _$50_ a month? 🤮

The Python lib actually has a [Google](https://python.langchain.com/en/latest/modules/agents/tools/examples/google_search.html) search wrapper, which is how I got the idea to leverage this custom search engine pattern.

They just added a [DuckDuckGo](https://python.langchain.com/en/latest/modules/agents/tools/examples/ddg.html) wrapper, which I've been using in my Python API. I'd love to dig into setting that up in JS, because there are no API keys to hassle with there.

Here is it in practice using the Langchain BabyAGI agent I setup with human in the loop feedback if you want to try it out

https://gist.github.com/mabry1985/cb36cb2a25d58628dcc2b506ec63e2dc

**Comment by shyagamzo** [comment]
Some code or a link would be useful :)

**Reply by mabry1985** [reply]
lol oops sorry

**Reply by mabry1985** [reply]
@shyagamzo updated to share code :smile:

**Comment by johnlindquist** [comment]
@mabry1985 @shyagamzo 

Some built-in inspiration you can steal from:

` from the main menu will run the "Google" script:
https://github.com/johnlindquist/kit/blob/main/src/main/google.ts

`~` from the main menu will run the "Suggest" script:
https://github.com/johnlindquist/kit/blob/main/src/main/suggest.ts

**Reply by mabry1985** [reply]
I'm still getting used to how the Gist/Discussion setup works here :sweat_smile: 

Yeah I saw the ` shortcut, that was where I was mentioning the redundancy, cool to see the code though! I've not seen that lib yet
